kernel/src/sched/sched_class/fair.rs

```diff
@@ -0,0 +1,294 @@
+    fn cmp(&self, other: &Self) -> cmp::Ordering {
+        self.key().cmp(&other.key())
+    }
+}
+
+/// The per-cpu run queue for the FAIR scheduling class.
+///
+/// See [`FairEntity`] for the explanation of vruntimes and scheduling periods.
+///
+/// The structure contains a `BTreeSet` to store the threads in the run queue to
+/// ensure the efficiency for finding next-to-run threads.
+pub(super) struct FairClassRq {
+    cpu: u32,
+    // FIXME: This field should have the type `BTreeSet<FairQueueItem>`. However,
+    // The `BTreeSet` implementation in the current Rust toolchain (2024-6-20) doesn't
+    // support cursors (e.g. `lower_bound_mut`), which is later merged into the Rust
+    // mainline in August, 2024. So please use `BTreeSet` after the toolchain update.
+    threads: BTreeMap<FairQueueItem, ()>,
+    load: u64,
+}
+
+impl core::fmt::Debug for FairClassRq {
+    fn fmt(&self, f: &mut core::fmt::Formatter<'_>) -> core::fmt::Result {
+        writeln!(
+            f,
+            "Fair: load = {}, num = {}",
+            self.load,
+            self.threads.len() + 1
+        )?;
+        writeln!(f, "  threads: ")?;
+        for thread in self.threads.keys() {
+            let vr = thread.key();
+            writeln!(f, "    {vr:?}")?;
+        }
+        Ok(())
+    }
+}
+
+impl FairClassRq {
+    pub fn new(cpu: u32) -> Self {
+        Self {
+            cpu,
+            threads: BTreeMap::new(),
+            load: 0,
+        }
+    }
+
+    /// The scheduling period is calculated as the maximum of the following two values:
+    ///
+    /// 1. The minimum period value, defined by [`min_period_clocks`].
+    /// 2. `period = min_granularity * n` where
+    ///    `min_granularity = log2(1 + num_cpus) * base_slice_clocks`, and `n` is the number of
+    ///    runnable threads (including the current running thread).
+    ///
+    /// The formula is chosen by 3 principles:
+    ///
+    /// 1. The scheduling period should reflect the running threads and CPUs;
+    /// 2. The scheduling period should not be too low to limit the overhead of context switching;
+    /// 3. The scheduling period should not be too high to ensure the scheduling latency
+    ///    & responsiveness.
+    fn period(&self) -> u64 {
+        let base_slice_clks = base_slice_clocks();
+        let min_period_clks = min_period_clocks();
+
+        let min_gran_clks = base_slice_clks * u64::from((1 + num_cpus()).ilog2());
+        // `+ 1` means including the current running thread.
+        (min_gran_clks * (self.threads.len() + 1) as u64).max(min_period_clks)
+    }
+
+    fn pop(&mut self, target_cpu: u32) -> Option<Arc<Thread>> {
+        let mut front = self.threads.lower_bound_mut(Bound::Unbounded);
+        let FairQueueItem(thread) = loop {
+            let (thread, _) = front.peek_next()?;
+            if thread.0.lock_cpu_affinity().contains(target_cpu) {
+                let (thread, _) = front.remove_next().unwrap();
```

Comment:
`target_cpu` was originally used for migration between run queues. In this case for now, it IS wrong because no migration is available, so I'll wrap it in an option to keep the behavior expected.